--- linux-2.6.30.9/include/linux/socket.h	2013-07-04 11:06:14.564664035 +0300
+++ linux-2.6.30.9/include/linux/socket.h	2013-07-04 11:05:14.634668886 +0300
@@ -31,7 +31,9 @@
 # endif
 #endif /* __KERNEL__ */
 
+#ifndef _LINUX_UN_H
 typedef unsigned short	sa_family_t;
+#endif
 
 /*
  *	1003.1g requires sa_family_t and that sa_data is char.
--- linux-2.6.30.9/include/linux/un.h	2013-07-04 11:06:38.397662106 +0300
+++ linux-2.6.30.9/include/linux/un.h	2013-07-04 11:06:24.052663267 +0300
@@ -2,7 +2,9 @@
 #define _LINUX_UN_H
 
 #define UNIX_PATH_MAX	108
+#ifndef _LINUX_SOCKET_H
 typedef unsigned short	sa_family_t;
+#endif
 struct sockaddr_un {
 	sa_family_t sun_family;	/* AF_UNIX */
 	char sun_path[UNIX_PATH_MAX];	/* pathname */
