Index: linux-2.6.30.9/arch/rlx/Makefile
===================================================================
--- linux-2.6.30.9.orig/arch/rlx/Makefile	2011-12-04 13:21:19.000000000 +0000
+++ linux-2.6.30.9/arch/rlx/Makefile	2011-12-04 13:21:25.000000000 +0000
@@ -32,6 +32,7 @@
 cflags-y := -ffunction-sections
 cflags-y += $(call cc-option, -mno-check-zero-division)
 cflags-y += -mabi=32
+cflags-y += -march=4181
 
 #
 # GCC uses -G 0 -mabicalls -fpic as default.  We don't want PIC in the kernel
