--- a/drivers/net/wireless/rtl8192cd/8192cd_cfg.h
+++ b/drivers/net/wireless/rtl8192cd/8192cd_cfg.h
@@ -1826,6 +1826,13 @@
 #undef CONFIG_RTK_VLAN_SUPPORT
 #endif
 
+#undef __DRAM_IN_865X
+#undef __IRAM_IN_865X
+#undef __MIPS16
+
+#define __DRAM_IN_865X  
+#define __IRAM_IN_865X  
+#define __MIPS16 
 
 #endif // _8192CD_CFG_H_
 
--- a/drivers/net/wireless/rtl8192cd/Makefile
+++ b/drivers/net/wireless/rtl8192cd/Makefile
@@ -4,6 +4,7 @@
 # Makefile for the Linux Wireless network device drivers.
 #
 
+EXTRA_CFLAGS += -I$(TOPDIR)/arch/rlx
 
 ifeq ($(CONFIG_RTL_92C_SUPPORT),y)
 RTL_WLAN_DATA_DIR := data
--- a/drivers/net/wireless/rtl8192cd/8192cd_headers.h
+++ b/drivers/net/wireless/rtl8192cd/8192cd_headers.h
@@ -1454,7 +1454,7 @@
 #undef EXTERN
 
 
-#if 0
+#if 1
 
 /*-----------------------------------------------------------------------------
 								8192cd_net80211.c
