--- a/rtkload/Makefile	2013-02-05 01:54:54.283265028 +0200
+++ b/rtkload/Makefile	2013-02-06 02:52:53.538612082 +0200
@@ -1,7 +1,7 @@
 # The kernel tree you do builds in.
 # Uncomment if you're building for the emulator
 include ../.config
-include ../../.config	# to check toolchain CONFIG_RSDK_rsdk-1.5.4-5281-EB-2.6.30-0.9.30.3-uls-101110
+#include ../../.config	# to check toolchain CONFIG_RSDK_rsdk-1.5.4-5281-EB-2.6.30-0.9.30.3-uls-101110
 
 KERNEL_ROOT = ..
 
