Index: linux-2.6.30.9/arch/rlx/kernel/module.c
===================================================================
--- linux-2.6.30.9.orig/arch/rlx/kernel/module.c
+++ linux-2.6.30.9/arch/rlx/kernel/module.c
@@ -387,10 +387,11 @@ int module_finalize(const Elf_Ehdr *hdr,
 {
 	const Elf_Shdr *s;
 	char *secstrings = (void *)hdr + sechdrs[hdr->e_shstrndx].sh_offset;
+	char dbe_table[]="__dbe_table";
 
 	INIT_LIST_HEAD(&me->arch.dbe_list);
 	for (s = sechdrs; s < sechdrs + hdr->e_shnum; s++) {
-		if (strcmp("__dbe_table", secstrings + s->sh_name) != 0)
+		if (strcmp(dbe_table, secstrings + s->sh_name) != 0)
 			continue;
 		me->arch.dbe_start = (void *)s->sh_addr;
 		me->arch.dbe_end = (void *)s->sh_addr + s->sh_size;
