--- a/drivers/gpio/Makefile	2012-11-05 14:07:42.074262520 +0400
+++ b/drivers/gpio/Makefile	2012-11-05 14:08:18.327603410 +0400
@@ -10,5 +10,6 @@
 obj-$(CONFIG_GPIO_PCA953X)	+= pca953x.o
 obj-$(CONFIG_GPIO_PCF857X)	+= pcf857x.o
 obj-$(CONFIG_GPIO_TWL4030)	+= twl4030-gpio.o
-obj-$(CONFIG_GPIO_XILINX)	+= xilinx_gpio.o
+obj-$(CONFIG_GPIO_RTL)		+= rtl_gpio.o
 obj-$(CONFIG_GPIO_BT8XX)	+= bt8xxgpio.o
+
--- a/drivers/gpio/Kconfig	2012-11-05 14:07:47.534263615 +0400
+++ b/drivers/gpio/Kconfig	2012-11-05 14:07:58.770932541 +0400
@@ -73,6 +73,15 @@
 	help
 	  Say yes here to support the Xilinx FPGA GPIO device
 
+config GPIO_RTL
+        bool "Realtek RTL819x GPIO support"
+        help
+          Say yes here to support the Realtek GPIO found in
+	  RTL819x boards. Please note, that this is a hacky 
+	  backport
+	  
+
+
 comment "I2C GPIO expanders:"
 
 config GPIO_MAX732X
