--- a/drivers/char/Makefile	2012-11-05 14:22:46.404451333 +0400
+++ b/drivers/char/Makefile	2012-11-05 14:23:57.307799712 +0400
@@ -98,7 +98,7 @@
 obj-$(CONFIG_GPIO_VR41XX)	+= vr41xx_giu.o
 obj-$(CONFIG_GPIO_TB0219)	+= tb0219.o
 obj-$(CONFIG_TELCLOCK)		+= tlclk.o
-obj-$(CONFIG_RTL_819X)		+= rtl_gpio.o
+obj-$(CONFIG_RTL_GPIO)		+= rtl_gpio.o
 obj-$(CONFIG_RTL_NFBI_MDIO)	+= rtl_mdio/
 
 obj-$(CONFIG_MWAVE)		+= mwave/
