--- a/arch/rlx/include/asm/string.h	2013-02-04 03:15:32.687233694 +0200
+++ b/arch/rlx/include/asm/string.h	2013-02-06 00:51:37.319166238 +0200
@@ -90,7 +90,7 @@
 	"3:\t.set\tat\n\t"
 	".set\treorder"
 	: "=r" (__cs), "=r" (__ct), "=r" (__res)
-	: "0" (__cs), "1" (__ct),  "m" (*__cs), "m" (*__ct));
+	: "0" (__cs), "1" (__ct));
 
   return __res;
 }
@@ -123,7 +123,7 @@
 	".set\tat\n\t"
 	".set\treorder"
 	: "=r" (__cs), "=r" (__ct), "=r" (__count), "=r" (__res)
-	: "0" (__cs), "1" (__ct), "2" (__count), "m" (*__cs), "m" (*__ct));
+	: "0" (__cs), "1" (__ct), "2" (__count));
 
 	return __res;
 }
