--- files/arch/rlx/bsp_rtl8196c/Makefile	2013-05-02 01:47:48.000000000 +0300
+++ files/arch/rlx/bsp_rtl8196c/Makefile.new	2013-05-11 17:20:45.000000000 +0300
@@ -7,7 +7,7 @@
 #
 # RTL8196C
 #
-include $(DIR_LINUX)/.config
+include $(TOPDIR)/.config
 
 core-y   += arch/rlx/bsp/
 cflags-y += -Iinclude/asm-rlx -Iarch/rlx/bsp/
@@ -20,8 +20,8 @@
 obj-$(CONFIG_PCI) += pci.o
 
 EXTRA_AFLAGS := $(CFLAGS)
-EXTRA_CFLAGS += -I$(DIR_LINUX)/drivers/net/rtl819x/AsicDriver
-EXTRA_CFLAGS += -I$(DIR_LINUX)/drivers/net/rtl819x/common
+EXTRA_CFLAGS += -I$(TOPDIR)/drivers/net/rtl819x/AsicDriver
+EXTRA_CFLAGS += -I$(TOPDIR)/drivers/net/rtl819x/common
 ifdef CONFIG_RTL_819X_SWCORE
 EXTRA_CFLAGS += -DCONFIG_RTL_819X_SWCORE
 endif
