diff --git a/target/linux/realtek/files/drivers/net/wireless/rtl8192cd/8192cd_security.h b/target/linux/realtek/files/drivers/net/wireless/rtl8192cd/8192cd_security.h
index a03656b..3212cc1 100644
--- a/drivers/net/wireless/rtl8192cd/8192cd_security.h
+++ b/drivers/net/wireless/rtl8192cd/8192cd_security.h
@@ -199,15 +199,6 @@ typedef enum{
 		DOT11_EVENT_WSC_PUTWLREQ_PROBEIND = 88,
 		DOT11_EVENT_WSC_PUTWLREQ_STATUSIND = 89,
 #endif
-#ifdef CONFIG_RTL_COMAPI_CFGFILE
-	DOT11_EVENT_WSC_START_IND = 90,
-	//EV_MODE, EV_STATUS, EV_MEHOD, EV_STEP, EV_OOB
-	DOT11_EVENT_WSC_MODE_IND = 91,
-	DOT11_EVENT_WSC_STATUS_IND = 92,
-	DOT11_EVENT_WSC_METHOD_IND = 93,
-	DOT11_EVENT_WSC_STEP_IND = 94,
-	DOT11_EVENT_WSC_OOB_IND = 95,
-#endif
 
 #ifdef P2P_SUPPORT
 		DOT11_EVENT_WSC_SWITCH_MODE = 100,
